### ISSUE TITLE
Support for querying nested objects with JDBC output

### ISSUE THREAD

**Issue by rli** [opendistro-for-elasticsearch/sql#1042]
Potentially related to https://github.com/opendistro-for-elasticsearch/sql/issues/897

We use the Amazon Elasticsearch Service, but this is reproducible using the 1.12.0 Docker Hub image

Given document:
```
POST /indexname/_doc/
{
    "timestamp": "1612990928597",
    "value": "0.0",
    "nested.metricName": "session_start"
}
```

The following SQL query with CSV output performs as expected:
```
POST _opendistro/_sql?format=csv
{
  "query": "SELECT nested.metricName FROM indexname;"
}

nested.metricName
session_start
```

However, the same query with JDBC output returns null values:
```
POST _opendistro/_sql?format=jdbc
{
  "query": "SELECT nested.metricName FROM indexname;"
}

{
  "schema": [{
    "name": "nested.metricName",
    "type": "text"
  }],
  "total": 1,
  "datarows": [[null]],
  "size": 1,
  "status": 200
}
```

If the field happens to be a keyword, casting the field as part of the query returns data rows as expected:
```
PUT _template/indexname2
{
  "index_patterns": "indexname2",
  "mappings": {
    "dynamic_templates": [
      {
        "metadata_strings": {
          "path_match": "nested.*",
          "match_mapping_type": "string",
          "mapping": {
            "type": "keyword"
          }
        }
      }
    ]
  }
}
...
...
POST /indexname2/_doc/
{
    "timestamp": "1612990928597",
    "value": "0.0",
    "nested.metricName": "session_start"
}
...
...
POST _opendistro/_sql?format=jdbc
{
  "query": "SELECT CAST(nested.metricName as STRING) FROM indexname2;"
}

{
  "schema": [{
    "name": "cast_nested.metricName",
    "type": "text"
  }],
  "total": 1,
  "datarows": [["session_start"]],
  "size": 1,
  "status": 200
}
```

However, this is not ideal as the query will never work if any document is missing a value for the requested field:
```
POST /indexname2/_doc/
{
    "timestamp": "1612990928597",
    "value": "0.0",
    "nested.metricName": "session_start",
    "nested.optionalValue": "value"
}
...
...
POST _opendistro/_sql?format=jdbc
{
  "query": "SELECT CAST(nested.optionalValue as STRING) FROM indexname2;"
}

{
  "error": {
    "reason": "Error occurred in Elasticsearch engine: all shards failed",
    "details": """Shard[0]: ScriptException[runtime error]; nested: IllegalStateException[A document doesn't have a value for a field! Use doc[<field>].size()==0 to check if a document is missing a field!];

For more details, please send request for Json format to see the raw response from elasticsearch engine.""",
    "type": "SearchPhaseExecutionException"
  },
  "status": 400
}
```

**Comment by penghuo** [comment]
Hey Richard,

Thanks for bing this issue. 

### TL;DR
We don't recomend to use "." inside the field name. In the following section, I will explain our design decision and explain why, and we do want to hear your use cases. 
**Note**: All the discussion below is based on our [new engine](https://github.com/opendistro-for-elasticsearch/sql/blob/develop/docs/dev/NewSQLEngine.md) will be released with upcoming ODFE 1.13. You can try it now with our [latest develop branch](https://github.com/opendistro-for-elasticsearch/sql) also.

### Explain
1. nested.metricName is not valid column name in SQL Spec.  [Reference](https://www.postgresql.org/docs/9.5/sql-expressions.html).
> 4.2.1. Column References
> A column can be referenced in the form:
> 
> correlation.columnname
> correlation is the name of a table (possibly qualified with a schema name), or an alias for a table defined by means of a FROM clause. The correlation name and separating dot can be omitted if the column name is unique across all the tables being used in the current query. (See also Chapter 7.)
2. the workaround is using **[quotes](https://github.com/opendistro-for-elasticsearch/sql/blob/develop/docs/user/general/identifiers.rst#delimited-identifiers)**.  for example, ```SELECT `nested.metricName` FROM indexname ```

```
POST /_opendistro/_sql?format=jdbc
{
  "query": """
    SELECT `nested.metricName` FROM indexname
  """
}
{
  "schema": [
    {
      "name": "nested.metricName",
      "type": "text"
    }
  ],
  "datarows": [
    [
      "session_start"
    ]
  ],
  "total": 1,
  "size": 1,
  "status": 200
}
```
3. Why we don't recomend to use "." inside the field name?
Becasuse, the field contain dot will making the index mapping conflict with data. Taking your sample data as example, the flattend mapping between field name and type will as follows. The "nested" field is exist in the index mapping, but it doesn't exist in the source data.
```
{
  "nested": "object"
  "nested.metricName": "text"
  "timestamp": "text"
 "value": text"
}
```
Considering the following query and result, the query is totally valid, because nested is a valid field in the index mapping, but no nested field exist in the source doc. so the result value is null.
```
SELECT nested FROM indexname

{
  "schema": [{
    "name": "nested",
    "type": "object"
  }],
  "total": 1,
  "datarows": [
    [null]
  ],
  "size": 1,
  "status": 200
}
```
4. The suggested solution is to keep your field without dot and using json nested structure. for example,
```
POST /index2/_doc/
{
    "timestamp": "1612990928597",
    "value": "0.0",
    "nested": {
     "metricName": "session_start" 
    }
}

POST /_opendistro/_sql?format=jdbc
{
  "query": """
    SELECT nested, nested.metricName FROM index2
  """
}

{
  "schema": [
    {
      "name": "nested",
      "type": "object"
    },
    {
      "name": "nested.metricName",
      "type": "text"
    }
  ],
  "total": 1,
  "datarows": [[
    {"metricName": "session_start"},
    "session_start"
  ]],
  "size": 1,
  "status": 200
}
```

**Comment by rli** [comment]
Thanks for the detailed response!

I can confirm that the new SQL engine works fine on a local cluster, but we'll need to wait for 1.13 to rollout to Amazon ES before we can try it out on real data.

Good point about preferring the object notation over dot notation. We've been explicitly flattening our documents without issue, but it's probably better to be explicit.
